Update e2e test for multi-entity saving in site editor

Addison-Stavlo opened the pull request:
## Description
<!-- Please describe what you have changed or added -->
Fixes skipped e2e tests regarding multi-entity saving in the site editor as added in: https://github.com/WordPress/gutenberg/pull/21159#issuecomment-607725131

Things Changed:
* Removed 1 assertion.  This was an inaccurate assertion on its own and was failing as a result.  
* Insured the test uses the `front-page` demo template as the template to make changes to.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->
Running e2e tests for multi-entity-saving locally and via travis CI.

## Types of changes
<!-- What types of changes does your code introduce?  -->
Bug fix (non-breaking change which fixes an issue)
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

Addison-Stavlo committed:
forced navigation to demo template

Addison-Stavlo commented:
@youknowriad - does this fix the site editor tests on your end by chance?

If not, do you have any more information about which tests are failing or your setup (enabled themes, available templates, which template is being loaded in the test, or anything else that might help me reproduce the issue on my end.)

youknowriad commented:
@Addison-Stavlo It was not just me, they were failing on travis. (the first one)

Addison-Stavlo commented:
Oh interesting.  Well this wouldn't fix the first one if thats the case, but I think ensuring we are on the correct template to test will be good nevertheless.

The first one asserts the save button is disabled when loading the site editor.  I guess this isn't always the case, I noticed a few times myself loading the Site Editor with the save button enabled.  But ever since saving that, it has subsequently always loaded disabled.  Perhaps this assertion doesn't make sense here... 🤔 we could probably start with the next assertions of it being enabled after edits are made and going from there..

Addison-Stavlo committed:
removed unwanted assertion, refactored selectors

youknowriad commented:
>  I think ensuring we are on the correct template to test will be good nevertheless.

We do have a posts reset each time you run e2e tests, we need something similar for templates I believe.

youknowriad commented:
>  I think ensuring we are on the correct template to test will be good nevertheless.

We do have a posts reset each time you run e2e tests, we need something similar for templates I believe.

Addison-Stavlo commented:
I removed the first test for now, I don't think that was an accurate assertion to begin with and thats why it was failing.

I added that extra step for ensuring we visit the demo template and did a small amount of refactoring.

Those tests are passing, there is some test failing but it's regarding `Editing modes (visual/HTML)`.

youknowriad commented on packages/e2e-tests/specs/experiments/multi-entity-saving.test.js:
```diff
@@ -20,14 +20,30 @@ import {
+	const templatePartSelector = '*[data-type="core/template-part"]';
+	const activatedTemplatePartSelector = `${ templatePartSelector } .block-editor-inner-blocks`;
+
+	// Re-usable assertiosn across Post/Site editors.
```
```suggestion
	// Re-usable assertions across Post/Site editors.
```

youknowriad commented:
Now it works on travis but the test fails locally as I believe it's affected by the theme choices, the template... Is there something we can do to ensure a consistent behavior. Maybe some resets in `setup-test-framework.js` like the `trashExistingPosts` function? I'm afraid the maintenance of this test become too hard.

Addison-Stavlo committed:
Update packages/e2e-tests/specs/experiments/multi-entity-saving.test.js

Co-Authored-By: Riad Benguella <benguella@gmail.com>

Addison-Stavlo commented:
> Maybe some resets in setup-test-framework.js like the trashExistingPosts function?

Yeah, we can do something similar to trash all templates / template parts.  However, template parts inherited by the theme (like the header) still show up in the site editor after doing so.  Perhaps we should enable a specific theme for the test, like "Twenty twenty"... or could that cause some issues elsewhere? 🤔

Addison-Stavlo committed:
added template cleanup and changed dirtying method

Addison-Stavlo committed:
Merge branch 'update/e2e-multi-entity-saving-site-editor' of https://github.com/WordPress/gutenberg into update/e2e-multi-entity-saving-site-editor

Addison-Stavlo commented:
A few updates:
* Updated `trashExistingPosts` function to be reusable for other post types (i.e. `wp_template` and `wp_template_part`.  I applied this cleanup to the `beforeAll` in our multi-site-editing file.  For now I didn't think it made sense to add them to the global suite cleanup where the current `trashExistingPosts` is called, it seems like unnecessary overhead for non site-editor related tests.

*  Updated the selector used in `open-global-block-inserter` to include the site editor, so that utils like `insertBlock` will work in site editor testing.

*  Changed how we are dirtying the entities in the site editor.  Instead of selecting the header and adding text, we are adding a template part.  This should be more universal and may eliminate the need to switch to a specific theme for congruency.  

It seems however, after a template part is inserted into the site editor and saved, trashing the template part from the template parts menu does not delete it from the template it was inserted to in the site editor.   Then saving from the site editor will un-trash them.  So the demo template on our test site is going to keep growing until this is solved or we find a better method 🤔 ... will keep thinking about this, but open to other ideas.  targeting it specifically to delete via UI in the site editor seems a bit tricky.

youknowriad reviewed (approved):
A lot better, thanks for the updates.

Addison-Stavlo committed:
added wait timeout after inserting template part in site editor

Addison-Stavlo commented:
Darn, I notice that in the non-headless version we are leaving placeholder template parts in our wake:

<img src=''https://user-images.githubusercontent.com/28742426/78609184-2dc6fb80-7830-11ea-8fd4-4b6fc06e0787.png" width="300px" />

![Screen Shot 2020-04-06 at 1 55 23 PM](https://user-images.githubusercontent.com/28742426/78609184-2dc6fb80-7830-11ea-8fd4-4b6fc06e0787.png)

It seems to be saving before the newly created template part is fully inserted.  I have added a short promise timeout to wait for one second after inserting the template part.  Generally, I wouldn't want to set a pause for an arbitrary amount of time, but am unable to find a selector to wait for that is specific enough in the context of the site editor.

@youknowriad  do you think the promise timeout is reasonable for now?  Alternatively we could try to do a cleanup of these uninitiated template parts `afterAll`, but that would be a bit uglier and this timeout should prevent them from remaining after the test in the first place.  🤔

youknowriad commented on packages/e2e-tests/specs/experiments/multi-entity-saving.test.js:
```diff
@@ -175,6 +175,11 @@ describe( 'Multi-entity save flow', () => {
 			await page.keyboard.press( 'Enter' );
 
+			// Wait to ensure template part is created/inserted before saving.
+			await new Promise( ( resolve ) => {
```
Its there a way to replace this with `waitForSelector` or something like that. We avoid timeouts and things like that in tests because these are very fragile.

youknowriad replied:
Commented too soon, just saw your comment above. I'll think about it but yeah we'll have to find a more reliable way.

Addison-Stavlo replied:
Yeah 🤔 , the selector for the inserted template part doesn't seem distinguishable from other template parts that may be present.  I wonder if it would make sense to add the created template's slug as part of a classname here?

Addison-Stavlo committed:
added site editor specific cleanup instead of timeout

Addison-Stavlo commented:
@youknowriad  - I think I figured a solution to getting around this template placeholder / timeout issue.  I wonder what you think of this approach?  ( 🤞 as Travis processes this, but I think it should be good ) 

I updated this so that in the test for the site editor, we only insert the Template Part placeholder block and do not initialize it.  Then added a site editor specific `afterAll` to delete that placeholder and re-save, so we should have no lingering artifacts from our test in the site editor.

While the post editor requires us to fully insert and dirty the template to test the multi-entity saving functionality, in the site editor we only need to dirty the top-level to test this.  Adding the placeholder achieves this and is easily targeted for cleanup (a seemingly trickier process than when we aren't able to just delete the top-level entity in cleanup like with the post editor).

youknowriad commented on packages/e2e-tests/specs/experiments/multi-entity-saving.test.js:
```diff
@@ -153,6 +157,28 @@ describe( 'Multi-entity save flow', () => {
+				entitiesSaveSelector
+			);
+			await saveButton.click();
+		} );
```
so if I understand properly, previously executed tests leave templates and template parts in the database that are used when we access the test again right? can't we just remove them by going to the templates listing under "Appearance"?

Addison-Stavlo replied:
> so if I understand properly, previously executed tests leave templates and template parts in the database that are used when we access the test again right? can't we just remove them by going to the templates listing under "Appearance"?

We are doing that as part of the `afterAll` cleanup here:  going to the template and template part listings under 'appearance' and removing them all.

Inside the Site Editor, if we have added a template part and then delete that template part from the listings in "appearance", it will not remove that template part from the site entity.  The site editor will load up with that template part in place still, and if you re-save everything it will re-save into 'appearance' as well.

That is why we are currently avoiding fully inserting fully template parts with edits to save in the site editor test.  Also, we don't need to do those things.  So now we only add the Template Part placeholder which is enough to dirty the site entity to allow us to trigger the save functionality, then we remove the placeholder as cleanup.

youknowriad replied:
>  If we were to move this to core-data, these would go along with it and Im assuming that would be fine? Or is there some other issue here?

If we remove both the templates and the template parts, won't we come back to the starting state (just theme files)?

Addison-Stavlo replied:
I thought that would be the case as well, but not completely?  Try going into the site editor, adding a template part, adding some content to it, and saving.  Then goto appearance and delete all your template parts.  Return to the site editor and notice the site editor still knows the template parts slugs and content, and is ready to save them all since it cannot find them in the database anymore.

youknowriad replied:
> Then goto appearance and delete all your template parts

I'm saying delete not just the template parts but also the templates, did you try that?

Addison-Stavlo replied:
Oh, yes.. thank you!  I thought I was removing templates as well, but it turns out `trashExistingPosts` wasn't functioning properly for template and template parts.   It was adding the query arg as `postType` as opposed to `post_type`.  So now we have an underscore in a key name in JavaScript to fix that 🤣 

```	
	const query = addQueryArgs( '', {
		post_type: postType,
	} ).slice( 1 );
```
A bit against standard JS convention, but it gets us to the right page now.  Do you see any potential issues with having the underscore here?

Now this cleanup is working as expected, so I removed that unnecessary cleanup function.

youknowriad replied:
underscore is ok there.

Addison-Stavlo committed:
removed cleanup, fixed trash post function

Addison-Stavlo committed:
whoops, delete the commented out goo

youknowriad commented:
This is way better now 👍

Addison-Stavlo commented:
> This is way better now 👍

Lol, seriously thank you for your help through this.  I watched the cleanup in the slow-mo headed mode and thought it was doing what it was supposed to, and I totally got off track when cleanup wasn't working as expected.
